src/SDL_log.c

```diff
@@ -239,85 +342,86 @@ static bool SDL_ParseLogCategoryPriority(const char *hint, int category, SDL_Log
             ++next;
         }
 
-        if (SDL_ParseLogCategory(name, (sep - name), &current_category)) {
-            if (current_category == category) {
-                const char *value = sep + 1;
-                size_t len;
-                if (next) {
-                    len = (next - value - 1);
+        if (ParseLogCategory(name, (sep - name), &category)) {
+            const char *value = sep + 1;
+            size_t len;
+            if (next) {
+                len = (next - value - 1);
+            } else {
+                len = SDL_strlen(value);
+            }
+            if (ParseLogPriority(value, len, &priority)) {
+                if (category == DEFAULT_CATEGORY) {
+                    for (int i = 0; i < SDL_arraysize(SDL_log_priorities); ++i) {
+                        if (SDL_log_priorities[i] == SDL_LOG_PRIORITY_INVALID) {
+                            SDL_log_priorities[i] = priority;
+                        }
+                    }
+                    SDL_log_default_priority = priority;
                 } else {
-                    len = SDL_strlen(value);
+                    SDL_SetLogPriority(category, priority);
                 }
-                return SDL_ParseLogPriority(value, len, priority);
             }
         }
     }
-    return false;
 }
 
-static SDL_LogPriority SDL_GetDefaultLogPriority(int category)
+void SDL_ResetLogPriorities(void)
 {
-    const char *hint = SDL_GetHint(SDL_HINT_LOGGING);
-    if (hint) {
-        SDL_LogPriority priority;
-
-        if (SDL_ParseLogCategoryPriority(hint, category, &priority)) {
-            return priority;
-        }
-        if (SDL_ParseLogCategoryPriority(hint, DEFAULT_CATEGORY, &priority)) {
-            return priority;
-        }
-    }
+    SDL_CheckInitLog();
 
-    switch (category) {
-    case SDL_LOG_CATEGORY_APPLICATION:
-        return SDL_LOG_PRIORITY_INFO;
-    case SDL_LOG_CATEGORY_ASSERT:
-        return SDL_LOG_PRIORITY_WARN;
-    case SDL_LOG_CATEGORY_TEST:
-        return SDL_LOG_PRIORITY_VERBOSE;
-    default:
-        return SDL_LOG_PRIORITY_ERROR;
-    }
-}
-
-SDL_LogPriority SDL_GetLogPriority(int category)
-{
-    SDL_LogLevel *entry;
+    SDL_LockMutex(SDL_log_lock);
+    {
+        CleanupLogPriorities();
 
-    for (entry = SDL_loglevels; entry; entry = entry->next) {
-        if (entry->category == category) {
-            return entry->priority;
+        SDL_log_default_priority = SDL_LOG_PRIORITY_INVALID;
+        for (int i = 0; i < SDL_arraysize(SDL_log_priorities); ++i) {
+            SDL_log_priorities[i] = SDL_LOG_PRIORITY_INVALID;
         }
-    }
 
-    if (SDL_forced_priority) {
-        return SDL_forced_priority_level;
-    }
-
-    return SDL_GetDefaultLogPriority(category);
-}
+        const char *hint = SDL_GetHint(SDL_HINT_LOGGING);
```

Comment:
Wasn't the concern in #9679 that SDL_GetHint calls SDL_getenv() and that was super-expensive? Or was that unrelated?

Reply:
Yes, the problem was that we were calling it on every call to SDL_GetLogPriority(). Now we only do it at startup or when the hint changes.